File: app/src/main/java/com/androidessence/cashcaretaker/dataTransferObjects/TransactionDetails.java
package com.androidessence.cashcaretaker.dataTransferObjects;

import android.database.Cursor;
import android.os.Parcel;

import com.androidessence.cashcaretaker.data.CCContract;

/**
 * Details about a transaction beyond the required info.
 *
 * Created by adam.mcneilly on 9/7/16.
 */
public class TransactionDetails extends Transaction {
    private Category category;

    public TransactionDetails(Cursor cursor) {
        super(cursor);
        // Get category name
        String categoryName = cursor.getString(cursor.getColumnIndex(CCContract.CategoryEntry.COLUMN_DESCRIPTION));
        setCategory(new Category(getCategoryID(), categoryName));
    }

    public TransactionDetails(Parcel source) {
        super(source);
        setCategory((Category) source.readParcelable(Category.class.getClassLoader()));
    }

    public static final Creator<TransactionDetails> CREATOR = new Creator<TransactionDetails>() {
        @Override
        public TransactionDetails createFromParcel(Parcel source) {
            return new TransactionDetails(source);
        }

        @Override
        public TransactionDetails[] newArray(int size) {
            return new TransactionDetails[size];
        }
    };

    public Category getCategory() {
        return category;
    }

    public void setCategory(Category category) {
        this.category = category;
    }

    public static Creator<TransactionDetails> getCREATOR() {
        return CREATOR;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeParcelable(getCategory(), flags);
    }
}

File: app/src/main/java/com/androidessence/cashcaretaker/dataTransferObjects/TransactionDetails.kt
package com.androidessence.cashcaretaker.dataTransferObjects

import android.database.Cursor
import android.os.Parcel
import android.os.Parcelable
import com.androidessence.cashcaretaker.creator

import com.androidessence.cashcaretaker.data.CCContract

/**
 * Details about a transaction beyond the required info.

 * Created by adam.mcneilly on 9/7/16.
 */
class TransactionDetails : Transaction {
    var category: Category? = null

    constructor(cursor: Cursor) : super(cursor) {
        // Get category name
        val categoryName = cursor.getString(cursor.getColumnIndex(CCContract.CategoryEntry.COLUMN_DESCRIPTION))
        category = Category(categoryID, categoryName)
    }

    constructor(source: Parcel) : super(source) {
        category = source.readParcelable<Parcelable>(Category::class.java.classLoader) as Category
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeParcelable(category, flags)
    }

    companion object {
        @JvmField val CREATOR = creator(::TransactionDetails)
    }
}
